crates/typst-layout/src/shapes.rs

```diff
@@ -90,28 +79,164 @@ pub fn layout_path(
+                        if path_closed && last != start {
+                            path.close_path();
+                        }
+                        let mut p = resolve(*point);
+                        if *rel {
+                            p += last;
+                        }
+                        path.move_to(p);
+                        size.x.set_max(p.x);
+                        size.y.set_max(p.y);
+                        last = p;
+                        last_control = p;
+                        start = p;
+                    }
+                    LineTo(point, rel) => {
+                        let mut p = resolve(*point);
+                        if *rel {
+                            p += last;
+                        }
+                        if path.is_empty() {
+                            path.move_to(start);
+                        }
+                        path.line_to(p);
+                        size.x.set_max(p.x);
+                        size.y.set_max(p.y);
+                        last = p;
+                        last_control = p;
+                    }
+                    QuadraticTo(end, c, rel) => {
+                        let mut end = resolve(*end);
+                        if *rel {
+                            end += last;
+                        }
+                        let c = match c {
+                            Some(Smart::Custom(c)) => {
+                                let mut c = resolve(*c);
+                                if *rel {
+                                    c += last;
+                                }
+                                c
+                            }
+                            Some(Smart::Auto) => 2. * last - last_control,
+                            None => end,
+                        };
+                        let c1 = last + (2. / 3.) * (c - last);
+                        let c2 = end + (2. / 3.) * (c - end);
+                        if path.is_empty() {
+                            path.move_to(start);
+                        }
+                        add_cubic(&mut path, &mut size, last, end, c1, c2);
+                        last = end;
+                        last_control = c;
+                    }
+                    CubicTo(end, c1, c2, rel) => {
+                        let mut end = resolve(*end);
+                        if *rel {
+                            end += last;
+                        }
+                        let c1 = match c1 {
+                            Some(Smart::Custom(c)) => {
+                                let mut c = resolve(*c);
+                                if *rel {
+                                    c += last;
+                                }
+                                c
+                            }
+                            Some(Smart::Auto) => 2. * last - last_control,
+                            None => last,
+                        };
```

Comment:
There is quite a bit of code duplication here. Overall, I feel like the function `layout_path` has gotten a bit too large here, and some factoring out is needed, either into subfunctions, or maybe into a struct with builder-style methods.